File: core/src/main/java/net/rickiekarp/core/net/ConnectionHandler.java
package net.rickiekarp.core.net;

import net.rickiekarp.core.AppContext;
import net.rickiekarp.core.debug.LogFileHandler;
import net.rickiekarp.core.settings.Configuration;
import okhttp3.*;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.net.ConnectException;
import java.net.URL;
import java.net.URLEncoder;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * A handler to add a layer of abstraction between network layer and implementation
 */
class ConnectionHandler {
	private static final String API_STRING							= "/api/";
	private static final MediaType MEDIA_TYPE_MARKDOWN              = MediaType.parse("application/json");
	private static final int MAX_CONNECTION_TIMEOUT_MILLISECONDS    = 15000;
	private static final int MAX_DATA_TRANSFER_TIMEOUT_MILLISECONDS = 40000;
	private static final int NO_PORT                                = -1;
	private static final String UTF_8                               = "UTF-8";
	private static final String HEADER_USER_AGENT                   = "User-Agent";
	private static final String HEADER_AUTHORIZATION                = "Authorization";
	private static final int RESPONSE_ERROR_CODE_SERVER_UNAVAILABLE = 503;
	private final OkHttpClient mHttpClient;

	ConnectionHandler() {
		mHttpClient = new OkHttpClient.Builder()
				.connectTimeout(MAX_CONNECTION_TIMEOUT_MILLISECONDS, TimeUnit.MILLISECONDS)
				.readTimeout(MAX_DATA_TRANSFER_TIMEOUT_MILLISECONDS, TimeUnit.MILLISECONDS)
				.build();
	}

	InputStream requestInputStream(final NetworkAction networkAction) {
		Response response = request(networkAction);
		if (response == null) {
			return null;
		}
		return response.body().byteStream();
	}

	Response request(final NetworkAction networkAction) {
		try {
			final URL hostUrl = new URL(Configuration.host + networkAction.getHostUrl());
			return performRequest(hostUrl, networkAction);
		} catch (final Exception e) {
			e.printStackTrace();
			return null;
		}
	}

	private Response performRequest(final URL hostUrl, final NetworkAction action) throws IOException {
		String decodedUrl = decodeUrlToString(hostUrl) + API_STRING + action.getActionUrl();

		final Request.Builder builder = new Request.Builder();

		switch (action.getMethod()) {
			case "GET":
				decodedUrl += encodeParametersToString(action.getParameterMap());
				builder.get();
				break;
			case "POST":
				final RequestBody requestBody = RequestBody.create(MEDIA_TYPE_MARKDOWN, encodeParametersToJson(action.getParameterMap()));
				builder.post(requestBody);
				break;
		}

		final URL composedUrl = new URL(decodedUrl);
		builder.url(composedUrl);

		addHeaders(action.getParameterMap(), builder);

		final Request request = builder.build();
		LogFileHandler.logger.info(request.method() + ": " + decodedUrl);
		//printRequestHeaders(request);
		//printRequestBody(request);

        try {
            final Response response = mHttpClient.newCall(request).execute();
            if (response.code() == RESPONSE_ERROR_CODE_SERVER_UNAVAILABLE) {
                throw new RuntimeException("Error");
            }
			//printResponseHeaders(response);
            return response;
        } catch (ConnectException e) {
            LogFileHandler.logger.severe(e.getMessage());
            return null;
        }
	}

	private void addHeaders(final Map<String, String> parameterMap, final Request.Builder builder) {
		builder.addHeader(HEADER_USER_AGENT, AppContext.getContext().getContextIdentifier() + "/" + AppContext.getContext().getInternalVersion());

		if (AppContext.getContext().getAccountManager().getAccount() != null) {
			if (AppContext.getContext().getAccountManager().getAccount().getAccessToken() != null) {
				builder.addHeader(HEADER_AUTHORIZATION, "Basic " + AppContext.getContext().getAccountManager().getAccount().getAccessToken());
			}
		}
	}

	/**
	 * Puts the parameters of the given parameter map into a JSON Object and returns its string representation
	 * @param params Parameter map to encode
	 * @return JSON String of the parameter map
	 */
	private String encodeParametersToJson(final Map<String, String> params) {
		JSONObject resultJson = new JSONObject();
		for (final Map.Entry<String, String> pair : params.entrySet()) {
			resultJson.put(pair.getKey(), pair.getValue());
		}
		return resultJson.toString();
	}

	/**
	 * Encodes the given parameter map to a URL parameter representation
	 * @param params Parameter map to encode
	 * @return URL representation of the parameter map
	 * @throws UnsupportedEncodingException UnsupportedEncodingException
	 */
	private String encodeParametersToString(final Map<String, String> params) throws UnsupportedEncodingException {
		final StringBuilder result = new StringBuilder();
		boolean isFirst = true;

		for (final Map.Entry<String, String> pair : params.entrySet()) {
			if (isFirst) {
				isFirst = false;
				result.append('?');
			} else {
				result.append('&');
			}

			result.append(URLEncoder.encode(pair.getKey(), UTF_8));
			result.append('=');
			result.append(URLEncoder.encode(pair.getValue(), UTF_8));
		}
		return result.toString();
	}

	private String decodeUrlToString(final URL url) {
		final String port;
		final int portInt = url.getPort();
		if (portInt != NO_PORT) {
			port = ":" + portInt;
		} else {
			port = "";
		}
		return url.getProtocol() + "://" + url.getHost() + port + url.getPath();
	}

	private String bodyToString(final Request request) {
//		try {
//			final Request copy = request.newBuilder().build();
//			final Buffer buffer = new Buffer();
//			copy.body().writeTo(buffer);
//			return buffer.readUtf8();
//		} catch (final IOException e) {
//			e.printStackTrace();
//			return null;
//		}
		return null;
	}

	private void printRequestBody(Request request) {
		if (request.method().equals("POST")) {
			System.out.println();
			System.out.println("Request body:");
			System.out.println(bodyToString(request));
			System.out.println();
		}
	}

	private void printRequestHeaders(Request request) {
		System.out.println();
		System.out.println("Request Headers:");
		for (int i = 0; i < request.headers().size(); i++) {
			System.out.println(request.headers().name(i) + ": " + request.headers().value(i));
		}
		System.out.println();
	}

	private void printResponseHeaders(Response response) {
		System.out.println();
		System.out.println("Response Headers:");
		for (int i = 0; i < response.headers().size(); i++) {
			System.out.println(response.headers().name(i) + ": " + response.headers().value(i));
		}
		System.out.println();
	}
}

File: projects/JavaFXApps/core/src/main/java/net/rickiekarp/core/net/ConnectionHandler.kt
package net.rickiekarp.core.net

import net.rickiekarp.core.AppContext
import net.rickiekarp.core.debug.LogFileHandler
import net.rickiekarp.core.settings.Configuration
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import org.json.JSONObject

import java.io.IOException
import java.io.InputStream
import java.io.UnsupportedEncodingException
import java.net.ConnectException
import java.net.URL
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * A handler to add a layer of abstraction between network layer and implementation
 */
internal class ConnectionHandler {
    private val mHttpClient: OkHttpClient

    init {
        mHttpClient = OkHttpClient.Builder()
                .connectTimeout(MAX_CONNECTION_TIMEOUT_MILLISECONDS.toLong(), TimeUnit.MILLISECONDS)
                .readTimeout(MAX_DATA_TRANSFER_TIMEOUT_MILLISECONDS.toLong(), TimeUnit.MILLISECONDS)
                .build()
    }

    fun requestInputStream(networkAction: NetworkAction): InputStream? {
        val response = request(networkAction) ?: return null
        return response.body!!.byteStream()
    }

    fun request(networkAction: NetworkAction): Response? {
        try {
            val hostUrl = URL(Configuration.host + networkAction.hostUrl)
            return performRequest(hostUrl, networkAction)
        } catch (e: Exception) {
            e.printStackTrace()
            return null
        }

    }

    @Throws(IOException::class)
    private fun performRequest(hostUrl: URL, action: NetworkAction): Response? {
        var decodedUrl = decodeUrlToString(hostUrl) + API_STRING + action.actionUrl

        val builder = Request.Builder()

        when (action.method) {
            "GET" -> {
                decodedUrl += encodeParametersToString(action.parameterMap)
                builder.get()
            }
            "POST" -> {
                val requestBody = RequestBody.create(MEDIA_TYPE_MARKDOWN, encodeParametersToJson(action.parameterMap))
                builder.post(requestBody)
            }
        }

        val composedUrl = URL(decodedUrl)
        builder.url(composedUrl)

        addHeaders(action.parameterMap, builder)

        val request = builder.build()
        LogFileHandler.logger.info(request.method + ": " + decodedUrl)
        //		printRequestHeaders(request);
        //		printRequestBody(request);

        try {
            val response = mHttpClient.newCall(request).execute()
            if (response.code == RESPONSE_ERROR_CODE_SERVER_UNAVAILABLE) {
                throw RuntimeException("Error")
            }
            //LogFileHandler.logger.info(String.valueOf(response.code()));
            //printResponseHeaders(response);
            return response
        } catch (e: ConnectException) {
            LogFileHandler.logger.severe(e.message)
            return null
        }

    }

    private fun addHeaders(parameterMap: Map<String, String>, builder: Request.Builder) {
        builder.addHeader(HEADER_USER_AGENT, AppContext.getContext().contextIdentifier + "/" + AppContext.getContext().internalVersion)

        if (AppContext.getContext().accountManager.account != null) {
            if (AppContext.getContext().accountManager.account!!.accessToken != null) {
                builder.addHeader(HEADER_AUTHORIZATION, "Basic " + AppContext.getContext().accountManager.account!!.accessToken!!)
            }
        }
    }

    /**
     * Puts the parameters of the given parameter map into a JSON Object and returns its string representation
     * @param params Parameter map to encode
     * @return JSON String of the parameter map
     */
    private fun encodeParametersToJson(params: Map<String, String>): String {
        val resultJson = JSONObject()
        for ((key, value) in params) {
            resultJson.put(key, value)
        }
        return resultJson.toString()
    }

    /**
     * Encodes the given parameter map to a URL parameter representation
     * @param params Parameter map to encode
     * @return URL representation of the parameter map
     * @throws UnsupportedEncodingException UnsupportedEncodingException
     */
    @Throws(UnsupportedEncodingException::class)
    private fun encodeParametersToString(params: Map<String, String>): String {
        val result = StringBuilder()
        var isFirst = true

        for ((key, value) in params) {
            if (isFirst) {
                isFirst = false
                result.append('?')
            } else {
                result.append('&')
            }

            result.append(URLEncoder.encode(key, UTF_8))
            result.append('=')
            result.append(URLEncoder.encode(value, UTF_8))
        }
        return result.toString()
    }

    private fun decodeUrlToString(url: URL): String {
        val port: String
        val portInt = url.port
        if (portInt != NO_PORT) {
            port = ":$portInt"
        } else {
            port = ""
        }
        return url.protocol + "://" + url.host + port + url.path
    }

    private fun bodyToString(request: Request): String? {
        //		try {
        //			final Request copy = request.newBuilder().build();
        //			final Buffer buffer = new Buffer();
        //			copy.body().writeTo(buffer);
        //			return buffer.readUtf8();
        //		} catch (final IOException e) {
        //			e.printStackTrace();
        //			return null;
        //		}
        return null
    }

    private fun printRequestBody(request: Request) {
        if (request.method == "POST") {
            println()
            println("Request body:")
            println(bodyToString(request))
            println()
        }
    }

    private fun printRequestHeaders(request: Request) {
        println()
        println("Request Headers:")
        for (i in 0 until request.headers.size) {
            println(request.headers.name(i) + ": " + request.headers.value(i))
        }
        println()
    }

    private fun printResponseHeaders(response: Response) {
        println()
        println("Response Headers:")
        for (i in 0 until response.headers.size) {
            println(response.headers.name(i) + ": " + response.headers.value(i))
        }
        println()
    }

    companion object {
        private val API_STRING = "/"
        private val MEDIA_TYPE_MARKDOWN = "application/json".toMediaTypeOrNull()
        private val MAX_CONNECTION_TIMEOUT_MILLISECONDS = 15000
        private val MAX_DATA_TRANSFER_TIMEOUT_MILLISECONDS = 40000
        private val NO_PORT = -1
        private val UTF_8 = "UTF-8"
        private val HEADER_USER_AGENT = "User-Agent"
        private val HEADER_AUTHORIZATION = "Authorization"
        private val RESPONSE_ERROR_CODE_SERVER_UNAVAILABLE = 503
    }
}
